### PR TITLE
Drain requests in native instead of managed

### PR DESCRIPTION
**Pull request by jkotalik**
For https://github.com/aspnet/AspNetCore/issues/6605 and https://github.com/aspnet/AspNetCore/issues/6555.

There is a race condition when shutting down and another request coming into the server. Managed perceives that there are no requests remaining, but in native, a request can be instantiated and try to call into managed, which causes app verifier to crash iisexpress/w3wp.

### CODE REVIEW

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp [hunk 1]
```diff
@@ -259,7 +265,7 @@ IN_PROCESS_APPLICATION::ExecuteApplication()
             if (m_waitForShutdown)
             {
                 const auto clrWaitResult = WaitForSingleObject(m_clrThread.native_handle(), m_pConfig->QueryShutdownTimeLimitInMS());
-                THROW_LAST_ERROR_IF(waitResult == WAIT_FAILED);
+                THROW_LAST_ERROR_IF(clrWaitResult == WAIT_FAILED);
```
Comment (jkotalik) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: This is unrelated, I just noticed that we were checking the wrong wait timeout here.

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp [hunk 2]
```diff
@@ -517,9 +524,23 @@ IN_PROCESS_APPLICATION::CreateHandler(
...
     return S_OK;
 }
+
+void
+IN_PROCESS_APPLICATION::HandleRequestCompletion()
+{
+    m_requestCount--;
+    if (m_fStopCalled && m_requestCount == 0)
+    {
+        if (m_pDrainRequestEvent != nullptr)
```
Comment (pakrym) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: Why would this ever be null?

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h [hunk 3]
```diff
@@ -62,6 +64,18 @@ class IN_PROCESS_APPLICATION : public InProcessApplicationBase
     StopIncomingRequests()
     {
         QueueStop();
+
+        LOG_INFOF(L"Waiting for requests to drain");
```
Comment (pakrym) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h]: Log count

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 4]
```diff
@@ -58,7 +69,20 @@ IN_PROCESS_HANDLER::AsyncCompletion(
     HRESULT     hrCompletionStatus
 )
 {
+    auto res = AsyncCompletionInternal(cbCompletion, hrCompletionStatus);
+    if (res != RQ_NOTIFICATION_PENDING)
+    {
+        m_pApplication->HandleRequestCompletion();
```
Comment (pakrym) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Why not in ctor and destructor?

**Comment by pakrym** [comment]
Add a test that opens a hanging connection and requests shutdown, assert that event logs are correct and it doesn't crash.

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp [hunk 5]
```diff
@@ -517,9 +524,23 @@ IN_PROCESS_APPLICATION::CreateHandler(
 {
     try
     {
+        m_requestCount++;
```
Comment (pakrym) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: Assert that you never increment 0 when m_fStopCalled is true

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp [hunk 6]
```diff
@@ -517,9 +524,23 @@ IN_PROCESS_APPLICATION::CreateHandler(
...
     CATCH_RETURN();
 
     return S_OK;
 }
+
+void
+IN_PROCESS_APPLICATION::HandleRequestCompletion()
+{
+    m_requestCount--;
+    if (m_fStopCalled && m_requestCount == 0)
```
Comment (pakrym) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: Assign result of `m_requestCount--;` to a var and condition on it

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h [hunk 7]
```diff
@@ -62,6 +64,18 @@ class IN_PROCESS_APPLICATION : public InProcessApplicationBase
     StopIncomingRequests()
```
Comment (pakrym) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h]: We were not blocking StopAsync before but are going to start now. Would it cause any issues? What if someone calls StopAsync on request thread?

Comment (jkotalik) [reply src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h]: I think it is a problem.

**Comment by jkotalik** [comment]
🆙 📅

File: src/Servers/IIS/IntegrationTesting.IIS/src/IISExpressDeployer.cs [hunk 8]
```diff
@@ -28,7 +28,7 @@ public class IISExpressDeployer : IISDeployerBase
         private const string FailedToInitializeBindingsMessage = "Failed to initialize site bindings";
         private const string UnableToStartIISExpressMessage = "Unable to start iisexpress.";
         private const int MaximumAttempts = 5;
```
Comment (pakrym) [inline src/Servers/IIS/IntegrationTesting.IIS/src/IISExpressDeployer.cs]: Huh?

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs [hunk 9]
```diff
@@ -6,6 +6,7 @@
 using System.IO;
 using System.Linq;
 using System.Net;
+using System.Net.Http;
```
Comment (pakrym) [inline src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs]: rem

File: src/Servers/IIS/IIS/samples/NativeIISSample/Startup.cs [hunk 10]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Linq;
+using System.Threading.Tasks;
```
Comment (pakrym) [inline src/Servers/IIS/IIS/samples/NativeIISSample/Startup.cs]: rem

File: src/Servers/IIS/IntegrationTesting.IIS/src/IISExpressDeployer.cs (hunk 8 above)
Comment (jkotalik) [reply src/Servers/IIS/IntegrationTesting.IIS/src/IISExpressDeployer.cs]: whoopz

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h [hunk 11]
```diff
@@ -154,12 +162,14 @@ class IN_PROCESS_APPLICATION : public InProcessApplicationBase
 
     PFN_ASYNC_COMPLETION_HANDLER    m_AsyncCompletionHandler;
     PFN_DISCONNECT_HANDLER          m_DisconnectHandler;
+    PFN_DRAIN_HANDLER               m_DrainHandler;
```
Comment (pakrym) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h]: RequestsDrainedHandler everywhere.

Comment (pakrym) [reply src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h]: Just Drain is a bit terse

File: src/Servers/IIS/IIS/src/Core/IISHttpServer.cs [hunk 12]
```diff
@@ -110,22 +113,11 @@ void RegisterCancelation()
...
-                {
-                    RegisterCancelation();
-                }
-                else
-                {
-                    // We have drained all requests. Block any callbacks into managed at this point.
-                    _nativeApplication.StopCallsIntoManaged();
-                    _shutdownSignal.TrySetResult(null);
-                }
+                RegisterCancelation();
```
Comment (pakrym) [inline src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: Do we need ` if (Interlocked.Exchange(ref _stopping, 1) == 1)` logic before that?

Comment (pakrym) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: We don't have test that uses cancellation token.

Comment (jkotalik) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: > Do we need if (Interlocked.Exchange(ref _stopping, 1) == 1) logic before that?

We don't, but it avoids calling into native again and calling into native after Dispose.

> We don't have test that uses cancellation token.

I'll add one. We didn't have a test that called StopAsync at all, so I'll add a second test where we use the cancellation token.

> Store the registration and dispose it in OnDrainComplete

 Yeah that looks like a mistake.

Comment (pakrym) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: There are too many places we call:

```
server._nativeApplication.StopCallsIntoManaged();
server._shutdownSignal.TrySetResult(null);
```

We should unify all of them and simplify this entire thing.

The problem with calling this early is that we would be still left with hanging request handlers so nothing would really stop.

Comment (pakrym) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: We shouldn't fix everything in this PR but let's not multiply duplicated code and understand exactly what we are doing.

Comment (jkotalik) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: > The problem with calling this early is that we would be still left with hanging request handlers so nothing would really stop.

Yeah you are definitely right. I think we should only call both in RequestDrainedHandler

Comment (jkotalik) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: Let's chat tomorrow about this. I think I resolved issues here (don't feel great about tests though).

**Review by pakrym — APPROVED** [review]

**Comment by pakrym** [comment]
Don't merge until preview 2  ships.

File: src/Servers/IIS/IIS/src/Core/IISHttpServer.cs [hunk 13]
```diff
@@ -97,6 +98,11 @@ public Task StopAsync(CancellationToken cancellationToken)
         {
             _nativeApplication.StopIncomingRequests();
 
+            _cancellationTokenRegistration = cancellationToken.Register(() =>
+            {
+                _shutdownSignal.TrySetResult(null);
```
Comment (davidfowl) [inline src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: Closure
